### PR TITLE
syntax - sh.vim - Unify bashStatement, kshStatement and shStatement

### PR DESCRIPTION
**Pull request by bfoersterling**
## Changes

Following the discussion on the mailing list.

I have looked at all keywords and checked if they are built into `bash`,
`dash`, `ksh`, or `sh`, and if they are external commands.\
The external programs are still in `shStatement` but they are visually \
separated by a new vimscript statement in a new line.\
Unfortunatelly there are a lot of differences between builtin commands \
especially between the different Kornshell flavors and also Dash misses a lot \
of builtins.\
Maybe the external commands can also be moved to their own syntax group?\
But the highlighting will probably stay the same so not sure if this is useful.

Restructured and simplified the logic of the `kornshell` flavors.
Tried to unify statements whenever possible.

Highlighting for the `rpm` command was removed as being too distribution specific.

Cleanup: formatted with `gg=G` and removed dead code.

## Rerecorded Tests

- `testdir/dumps/sh_09_00.dump` (`let` is not a keyword in `dash`)
- `testdir/dumps/sh_ksh2020_05.dump` (`rename` is an ext. command that may be highlighted)
- `testdir/dumps/sh_ksh93u_05.dump` (`rename` is an ext. command that may be highlighted)
- `testdir/dumps/sh_ksh93v_05.dump` (`rename` is an ext. command that may be highlighted)
- `testdir/dumps/sh_ksh88_05.dump` (`rename` is an ext. command that may be highlighted)
- `testdir/dumps/sh_sundrous_03.dump` (`xgrep` should not be highlighted in `bash`)

```
$ let foo='bar'
dash: 1: let: not found
```

`rename` usually comes preinstalled through https://github.com/util-linux/util-linux/tree/master.

`xgrep` is neither a builtin in bash nor an external command that I could find.

### CODE REVIEW

File: runtime/syntax/sh.vim [hunk 1]
```diff
@@ -463,76 +455,76 @@ syn match   shEscape	contained	'\%(^\)\@!\%(\\\\\)*\\.'	nextgroup=shComment
...
+    syn keyword bashStatement bind builtin caller compgen complete compopt declare dirs disown enable fg help history let logout mapfile popd pushd readarray shopt source suspend time
+    syn keyword bashStatement typeset nextgroup=shSetOption
+    syn keyword bashAdminStatement daemon reload restart start status stop
```
Comment (chrisbra) [inline runtime/syntax/sh.vim]: daemon is of little use, isn't it?

**Comment by chrisbra** [comment]
Thanks, I think this makes sense.

File: runtime/syntax/sh.vim [hunk 2]
```diff
@@ -797,49 +786,29 @@ syn region shParen matchgroup=shArithRegion start='\$\@!(\%(\ze[^(]\|$\)' end=')
 
 " Additional sh Keywords: {{{1
 " ===================
-syn keyword shStatement break cd chdir continue eval exec exit kill newgrp pwd read readonly return shift test trap ulimit umask wait
+" builtins:
+syn keyword shStatement alias bg break cd continue command eval exec exit export fc getopts hash jobs local printf read readonly return shift times trap type ulimit umask unalias wait
```
Comment (dkearns) [inline runtime/syntax/sh.vim]: I haven't tried but this should break the `shAlias` match as the `syn-keyword` will have priority.

**Comment by dkearns** [comment]
> `xgrep` is neither a builtin in bash nor an external command that I could find.

This is probably `grep -X` from AT&T's AST Toolkit.  It gets a mention in the `ksh` man page.

File: runtime/syntax/sh.vim (hunk 2 above)
Comment (bfoersterling) [reply runtime/syntax/sh.vim]: Yes, it actually broke it for some reason.

I changed it back and added the test `runtime/syntax/testdir/input/sh_bash_alias.sh`.\
Maybe you can check if it is correct now and if you have an edge case in mind add a test for it.

Thanks for the review.

File: runtime/syntax/sh.vim (hunk 1 above)
Comment (bfoersterling) [reply runtime/syntax/sh.vim]: Right, I have removed it in the latest commit.

`reload`, `restart`, `start`, `status`, and `stop` are not that useful either in my opinion, but maybe someone wants these for `sys5init` scripts.\
And I don't think these script have to be in `bash`, so it should rather be named `shAdminStatement` and put in global scope.

**Comment by bfoersterling** [comment]
> > `xgrep` is neither a builtin in bash nor an external command that I could find.
> 
> This is probably `grep -X` from AT&T's AST Toolkit. It gets a mention in the `ksh` man page.

I don't think this should be highlighted in Bash.

For the kornshell:\
According to the tests it is available in `ksh88`, `ksh93v`, `ksh93u` and `ksh2020`.\
I tested `mksh` from alpine's `apk` and `ksh` from Arch Linux' `pacman` (which seems to be `ksh93`) and there was no `xgrep` command.\
But I guess it is better to keep it for the `kornshell`s as I am not familiar with `kornshell`.

**Comment by chrisbra** [comment]
thanks, let me include it
